dataflow/snippets/src/test/java/com/example/dataflow/PubSubWriteIT.java

```diff
@@ -0,0 +1,113 @@
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.SubscriptionName;
+import com.google.pubsub.v1.TopicName;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class PubSubWriteIT {
+  private static final String projectId = System.getenv("GOOGLE_CLOUD_PROJECT");
+
+  private ByteArrayOutputStream bout;
+  private PrintStream out;
+  private String topicId;
+  private String subscriptionId;
+  TopicAdminClient topicAdminClient;
+  SubscriptionAdminClient subscriptionAdminClient;
+
+  @Before
+  public void setUp() throws IOException {
+    bout = new ByteArrayOutputStream();
+    out = new PrintStream(bout);
+    System.setOut(out);
+
+    topicId = "test_topic_" + UUID.randomUUID().toString().substring(0, 8);
+    subscriptionId = topicId + "-sub";
+
+    TopicName topicName = TopicName.of(projectId, topicId);
+    topicAdminClient = TopicAdminClient.create();
+    topicAdminClient.createTopic(topicName);
+
+    SubscriptionName subscriptionName = SubscriptionName.of(projectId, subscriptionId);
+    subscriptionAdminClient = SubscriptionAdminClient.create();
+    subscriptionAdminClient.createSubscription(subscriptionName, topicName,
+        PushConfig.getDefaultInstance(), 120);
+  }
+
+  @After
+  public void tearDown() {
+    subscriptionAdminClient.deleteSubscription(SubscriptionName.of(projectId, subscriptionId));
+    topicAdminClient.deleteTopic(TopicName.of(projectId, topicId));
+    System.setOut(null);
+  }
+
+  @Test
+  public void testPubSubWriteWithAttributes() throws Exception {
+
+    final AtomicInteger count = new AtomicInteger(0);
+
+    PubSubWriteWithAttributes.main(
+        new String[] {
+            "--runner=DirectRunner",
+            "--topic=" + String.format("projects/%s/topics/%s", projectId, topicId)
+        });
+
+    // Verify that the pipeline wrote messages to Pub/Sub
+    MessageReceiver receiver =
+        (PubsubMessage message, AckReplyConsumer consumer) -> {
+          count.incrementAndGet();
+          consumer.ack();
+        };
+
+    Subscriber subscriber = null;
+    try {
+      ProjectSubscriptionName subscriptionName =
+          ProjectSubscriptionName.of(projectId, subscriptionId);
+
+      subscriber = Subscriber.newBuilder(subscriptionName, receiver).build();
+      subscriber.startAsync().awaitRunning();
+      subscriber.awaitTerminated(30, TimeUnit.SECONDS);
+    } catch (TimeoutException timeoutException) {
+      subscriber.stopAsync();
+    }
+    assertEquals(4, count.get());
```

Comment:
AFAIK subscription can get more than one copy of the message at some events. I suggest to validate that the each of the sent messages is received.

Reply:
Thanks -- yes, that's correct. By default, it's possible to receive a message more than once. I updated the test to collect the messages, by unique message ID.